tests/test_fastapi_roles.py

```diff
@@ -0,0 +1,126 @@
+from typing import Awaitable, Callable, Tuple
+
+import pytest
+import sqlalchemy as sa
+from fastapi import FastAPI, HTTPException, Request, Response
+from fastapi.testclient import TestClient
+from starlette.middleware.base import BaseHTTPMiddleware
+
+# Public API
+from dbos import DBOS, DBOSContextEnsure
+
+# Private API because this is a unit test
+from dbos.context import assert_current_dbos_context
+from dbos.error import DBOSDuplicateWorkflowEventError, DBOSNotAuthorizedError
+
+
+def test_simple_endpoint(dbos_fastapi: Tuple[DBOS, FastAPI]) -> None:
+    dbos, app = dbos_fastapi
+
+    class SetRoleMiddleware(BaseHTTPMiddleware):
+        async def dispatch(
```

Comment:
Can we directly use the `@app.middleware()` decorator instead of this?

We also want to capture authenticated user/role in our traces, so we need something like:
```
        DBOS.span.set_attribute("authenticated_user", "user1")
```

Reply:
Yes, you can use @app.middleware if you want.  I changed the test to show this... 6 of one, half-dozen of the other so whichever you'd rather see.

I think an earlier version of this comment asked about non-`async`, which I believe is not possible without work; anything in the middleware will block the event loop unless we implement a way to use a thread pool from within the middleware.

I will find a way to test the span stuff next.

Reply:
I was wrong in the earlier comment so I deleted it. It's okay to call async code from sync but not the other way around. Because if you call sync code from async it will block everything...

Reply:
We should also set "assumed_role" and "authenticated_roles" besides authenticated users.

Reply:
Those fields ("assumed_role") etc. are in the context, regardless of whether they are in the span.
I have them in the span now, and tests for that.
Do you want the info added to system DB as part of this PR or a follow-on?

Reply:
As we discussed in the standup, let's do the system DB changes in the same PR.

Reply:
I'm on it.  Compiles / mypy is happy, but needs additional testing.

Reply:
Tested.